src/xrpld/overlay/detail/PeerImp.cpp

```diff
@@ -905,15 +915,19 @@ PeerImp::onReadMessage(error_code ec, std::size_t bytes_transferred)
 {
     if (!socket_.is_open())
         return;
-    if (ec == boost::asio::error::operation_aborted)
+
+    // we started closing the local connection, stop reading
+    if (ec == boost::asio::error::operation_aborted || shutdown_)
         return;
     if (ec == boost::asio::error::eof)
     {
+        // Peer initiated connection close, just clean up
         JLOG(journal_.info()) << "EOF";
-        return gracefulClose();
+        return close();
```

Comment:
I think `gracefulClose()` should still be called. `gracefulClose()` calls `async_shutdown`, which is graceful and secure connection shutdown, especially with SSL/TLS streams. Calling close() might be an ungraceful and often insecure way to terminate a connection.